Task REST API 작성

EunjinJwa opened the pull request:
안녕하세요!
기능 구현정도 작성한 상태이고 리뷰 요청드립니다.^^ 

Task 요청 외의 다른 리소스 요청도 있을것을 염두해고 TaskHttpHandler를 작성했는데요,
대부분의 요청에서 필요로 할 ObjectMapper 객체 접근 및 사용에 대해 궁금한 점이 있어요. 
ObjectMapper 생성비용을 줄이기 위해 별도의 `JsonObjectMapper` 클래스를 만들어 json <-> Object 변환을 하고자 했습니다.
근데.. JsonObjectMapper 는 공통으로 사용하는 개념의 util성 클래스인데, Task 클래스를 인자로 받아 json -> Class 로 변환해주는 부분이 있는게 깔끔하지 못한 구조인 것 같다는 생각이 듭니다ㅠ 

그럼 잘 부탁드립니다 :)

EunjinJwa committed:
Task REST API 작성

- GET /tasks
- GET /tasks/{id}
- POST /tasks
- PUT /tasks/{id}
- DELETE /tasks/{id}

hannut91 commented on app/src/main/java/com/codesoom/assignment/util/JsonObjectMapper.java:
```diff
@@ -0,0 +1,29 @@
+		return objectMapper.writeValueAsString(object);
+	}
+
+	public static Task toTask(String jsonString) throws JsonProcessingException {
+		return objectMapper.readValue(jsonString, Task.class);
+	}
```
JSON, Object 변환을 위한 유틸 함수를 만드셨군요. 그런데 Task를 변환하는 것은 구체적인 타입을 받아야해서 고민이셨던 것 같습니다. 
제너릭을 이용해서 변환할 타입을 밖에서 받으면 어떨까요? 아마도 이렇게 될 것 같습니다.

```java
public static <T> T toTask(String jsonString, Class<T> valueType) throws JsonProcessingException {
	return objectMapper.readValue(jsonString, valueType);
}
```

사용할때는 이렇게 사용합니다.
```java
Task task = JsonObjectMapper.toTask(requestBody, Task.class);
```


See also
  - https://docs.oracle.com/javase/tutorial/java/generics/methods.html

hannut91 commented on app/src/main/java/com/codesoom/assignment/httphandler/TaskHttpHandler.java:
```diff
@@ -0,0 +1,111 @@
+		return Integer.parseInt(split[split.length - 1]);
+	}
+
+	private Task findTask(int taskId) {
+		return tasks.stream().filter(task -> task.getId() == taskId).findFirst().orElse(null);
+	}
```
`tasks`에서 우리가 원하는 task를 찾지 못할 경우 `null`을 반환하고 있는 것 같아요.이 함수를 사용하는 입장에서는 이 함수가 `null`을 반환할 수 있는지 예상할 수 없고, 그리고 `null`처리를 하지 않아도 자바 컴파일러는 아무런 불만도 표현하지 않기 때문에 위험성이 있는 것 같아요. 만약 반환 타입을 Optional 그대로 반환한다면, 함수의 Signature에서 이 값이 있을 수도 있고 없을 수도 있다는 의도를 드러내고 있고, 없는 경우를 처리하지 않으면 컴파일도 되지 않아서 에러 처리를 강제할 수 있어요

하지만 Optional 반환은 다른 부분이 어떻게 만들어졌냐가 관건이라고 보는데, 함수형으로 처리 중이 아니라면 예외가 가장 일반적이죠. Optional을 리턴하면 사용하는 쪽에서 비슷한 코드를 작성해야 하기 때문에 사용하기 힘든 인터페이스가 될 가능성이 커요.

EunjinJwa replied:
TaskNotFoundException 를 생성하여 예외를 발생시키도록 했습니다. 
말씀주신 것처럼 null 처리보다 예외 처리가 훨씬 명확한 코드작성이 가능한 것 같아요.

hannut91 commented on app/src/main/java/com/codesoom/assignment/model/Task.java:
```diff
@@ -0,0 +1,39 @@
+	}
+
+	@Override
+	public String toString() {
+		return "Task{" +
+				"id=" + id +
+				", title='" + title + '\'' +
+				'}';
+	}
```
String.format을 사용해서 표현하는 건 어떨까요? 다음과 같이 표현할 수 있을거에요

```java
return String.format("{ id = %s, title = %s }", id, title);
```

더 좋은건 Template literal이지만 자바에서는 지원하질 않네요 ㅠ 만약 이런게 지원하는 코틀린을 사용했다면 다음과 같이 사용할 수 있을거에요

```java
return "{ id = $id, title = $title }"
```

EunjinJwa replied:
IDE 자동 생성 기능을 사용했는데^^;; String.format을 사용하여 원하는 포맷으로 지정될 수 있도록 수정했습니다.

hannut91 commented on app/src/main/java/com/codesoom/assignment/httphandler/TaskHttpHandler.java:
```diff
@@ -0,0 +1,111 @@
+	}
+
+	private int extractTaskIdFromPath(String path) {
+		String[] split = path.split("/");
```
Path에서 `/`로 구분한 문자열들을 저장하신 것 같아요. Path에서 `/`로 구분된 요소들을 Path segment라고 부릅니다. 해당 도메인에서 실제로 사용하는 이름을 사용하는게 좋을 것 같습니다.

> A path component, consisting of a sequence of path segments separated by a slash (/)

See also
  - https://en.wikipedia.org/wiki/URL

hannut91 commented on app/src/main/java/com/codesoom/assignment/httphandler/TaskHttpHandler.java:
```diff
@@ -0,0 +1,111 @@
+		String requestBody = parseRequestBody(httpExchange);
+
+		Task task = JsonObjectMapper.toTask(requestBody);
+		task.setId(++taskId);
```
전위연산자를 사용해서 값이 변경되기 전의 값을 할당하고 나서 값을 증가시키고 있는 것 같아요. 사용법을 정확히 알고 있다면 ++value 와 value++ 을 구분해서 사용할 수 있겠지만, 사실 조금 헷갈릴수도 있는 문법이에요. 우리가 한 번 더 생각해야되는거죠. 아주 조금이라도 헷갈릴 수 있거나 다르게 해석될 여지가 있다면 저는 사용을 지양하는 것이 좋다고 생각해요.

EunjinJwa replied:
IdGenerator 클래스를 통해 여러 종류의 Id값의 생성을 관리할 수 있도록 수정했습니다.

뭔가 Tasks 의 목록 및 생성을 관리하는 객체를 만들까도 고민했는데, 일단 TaskHandler에서 관련 처리를 하는 정도로 정리했어요.

hannut91 commented on app/src/main/java/com/codesoom/assignment/httphandler/TaskHttpHandler.java:
```diff
@@ -0,0 +1,111 @@
+		return bufferedReader.lines().collect(Collectors.joining("\n"));
+	}
+
+	private void response(HttpExchange httpExchange, int httpCode, String response) throws IOException {
+		OutputStream responseBody = httpExchange.getResponseBody();
+		if (response != null) {
+			httpExchange.sendResponseHeaders(httpCode, response.getBytes().length);
+			responseBody.write(response.getBytes());
+		} else {
+			httpExchange.sendResponseHeaders(httpCode, 0);
+		}
+		responseBody.close();
+	}
```
메서드명과 변수명이 똑같이 `response`네요. 이름이 헷갈릴 수 있겠어요.

그리고 만약 주어진 String response가 null이라면 `httpExchange.getResponseBody()`는 실행할 필요가 없으니, 코드의 순서를 조금 정렬할 필요가 있어보입니다

hannut91 commented on app/src/main/java/com/codesoom/assignment/httphandler/TaskHttpHandler.java:
```diff
@@ -0,0 +1,111 @@
+	private void deleteTask(HttpExchange httpExchange) throws IOException {
+		String path = httpExchange.getRequestURI().getPath();
+		int taskId = extractTaskIdFromPath(path);
+		Task task = findTask(taskId);
+
+		if (task != null) {
+			tasks.remove(task);
+		}
+		response(httpExchange, 200, null);
```
```suggestion
		Task task = findTask(taskId);
		if (task != null) {
			tasks.remove(task);
		}

		response(httpExchange, 200, null);
```

관련이 있는 코드들끼리는 가까이 위치하고, 관련이 없는 코드들은 빈 줄로 구분해주는 것이 좋습니다. 우리가 글을 읽을 때 문단이 적절히 나눠져있다면, 새로운 내용이 시작된다는 것을 알고 읽잖아요. 코드도 마찬가지입니다. 그래서 비슷한 맥락끼리 같이 읽을 수 있도록 적절히 묶어주고, 관련이 없는 코드들은 적절히 구분해주는게 좋습니다.

EunjinJwa committed:
제너릭을 사용한 JsonObjectMapper.toObject 메서드 생성

EunjinJwa committed:
Task.toString 반환을 String.format 함수 사용으로 변경

EunjinJwa committed:
code review 반영

- 변수 및 메서드 이름 변경(pathSegments, sendResponse)
- 가독성 및 관련 코드들의 구분을 위한 메서드 줄바꿈 처리
- IdGenerator 클래스를 사용하여 시퀀스한 ID 값 생성

EunjinJwa committed:
sendResponse 동작 버그 수정

EunjinJwa committed:
test 통과를 위한 httpCode, port=8000 수정

EunjinJwa committed:
test 통과를 위한 httpCode, port=8000 수정

EunjinJwa committed:
IdGenerator package 를 util로 변경

hannut91 commented on app/src/main/java/com/codesoom/assignment/httphandler/DemoHttpHandler.java:
```diff
@@ -0,0 +1,23 @@
+
+	@Override
+	public void handle(HttpExchange httpExchange) throws IOException {
+		String response = "It's OK";
+		httpExchange.sendResponseHeaders(200, response.getBytes().length);
+		OutputStream responseBody = httpExchange.getResponseBody();
+		responseBody.write(response.getBytes());
+		responseBody.close();
```
200이라는 Magic literal이 사용되고 있네요. HTTP Status 200 Ok를 쓰신 것 같네요. static final로 선언하거나 enum같은 것을 사용할 수도 있겠지만, 성공을 주는 객체를 만들수도 있을 것 같습니다.

```java
public void handle(HttpExchange httpExchange) throws IOException {
    new ResponseSuccess(exchange).send("It's OK");
}

// ResponsSuccess.java
class ResponseSuccess {
    private static final int STATUS_CODE = 200;

    private final HttpExchange exchange;

    ResponseSuccess(HttpExchange exchange) {
        this.exchange = exchange;
    }

    public void send(String content) throw IOException {
        httpExchange.sendResponseHeaders(200, content.getBytes().length);

        OutputStream outputStream = httpExchange.getResponseBody();

        outputStream.write(content.getBytes());
        outputStream.close();
    }
}
```

EunjinJwa replied:
Magic literal..! 넵 요런 표현은 지양해야겠어요! 
응답코드별 객체를 각각 생성하고 HttpResponse 객체를 상속받아서 send 처리를 호출할 수 있도록 하였습니다.

hannut91 commented on app/src/main/java/com/codesoom/assignment/httphandler/TaskHttpHandler.java:
```diff
@@ -0,0 +1,121 @@
+	}
+
+	private void deleteTask(HttpExchange httpExchange) throws IOException {
+		String path = httpExchange.getRequestURI().getPath();
+		int taskId = extractTaskIdFromPath(path);
```
요청한 path에서 id를 얻고 있는 것 같습니다. 메서드의 이름이 명확해서 무엇을 하는지 알기가 쉽네요. 하지만 어떻게 하는지 너무 많이 드러내서 이 메서드에서 몰라도될 정보들을 많이 알게된 것 같습니다.

다음 커멘트를 한 번 읽어보세요 https://github.com/CodeSoom/spring-week1-assignment-1/pull/140#discussion_r1185648184

hannut91 commented on app/src/main/java/com/codesoom/assignment/httphandler/TaskHttpHandler.java:
```diff
@@ -0,0 +1,121 @@
+		sendResponse(httpExchange, 201, JsonObjectMapper.toJson(task));
+	}
+
+	private int extractTaskIdFromPath(String path) {
+		String[] pathSegments = path.split("/");
+
+		return Integer.parseInt(pathSegments[pathSegments.length - 1]);
+	}
```
path에서 마지막 문자열을 읽어서 숫자로 변환해주고 있는 것 같습니다. 정규식을 이용해서 문제를 해결해 볼수도 있을 것 같아요

EunjinJwa committed:
Review  반영 #1

TaskHandler에서 Http요청 및 응답 처리에 필요한 로직을 분리
Request
- HttpRequest 를 생성하여 Request에 대한 공통 처리를 할 수 있는 부모 클래스
- HttpRequest 를 상속받는 HttpTaskRequest 객체를 생성하여 TaskHandler 에서 Task 관련 처리를 할 수 있도록 함(taskId 추출)
Response
- HttpStatusCode 단위의 개별 객체 생성하여 응답코드 정의
- HttpResponse 상위 클래스를 생성하여 send() 메서드의 공용 사용

EunjinJwa committed:
Http 요청 및 응답에 관한 처리와 Task 데이터 핸들링에 대한 처리를 분리함.

- TaskService 에서 Task 에 대한 CRUD 비즈니스 로직 전담.
- TaskHttpHandler는 요청에 의한 비즈니스로직 호출 및 응답을 전담하도록 함.

EunjinJwa committed:
줄바꿈 및 코드 정리

EunjinJwa committed:
Review 반영 2

- Path의 Id 추출방식을 정규식 활용

EunjinJwa commented:
### Review 내용 반영
- TaskHandler와 Request, Response 처리를 별개의 객체로 생성하여, TaskHandler에서 굳이 몰라서 되는 내용 및 Response 응답코드 관련 설정을 분리하였습니다.
- path에서 taskId 를 식별할 수 있도록 정규식을 사용하여 추출하였습니다. (URIPatternMatcher)

### 기타 
- TaskService 객체를 생성하여 TaskHandler와 Task 관련 비즈니스로직(CRUD) 에 대한 책임을 각각 담당하도록 했습니다.

EunjinJwa committed:
GET 요청 구분 로직 개선

- 특정 리소스 조회인지, 모든 리소스 요청인지에 대한 path 구분 로직을 정규식 사용하여 rootPath 만 요청하였는지를 확인하는 메소드를 통해 구분함.

hannut91 commented on app/src/main/java/com/codesoom/assignment/http/HttpResponse.java:
```diff
@@ -0,0 +1,36 @@
+		this.statusCode = statusCode;
+	}
+
+	public void send() throws IOException {
+		sendResponse();
+	}
+
+	public void send(String responseBody) throws IOException {
+		sendResponse(responseBody);
+	}
+
+	private void sendResponse(String responseBody) throws IOException {
+		httpExchange.sendResponseHeaders(statusCode, responseBody.getBytes().length);
+		httpExchange.getResponseBody().write(responseBody.getBytes());
+		httpExchange.getResponseBody().close();
+	}
+
+	private void sendResponse() throws IOException {
+		httpExchange.sendResponseHeaders(statusCode, 0);
+		httpExchange.close();
+	}
```
sendResponse라는 프라이빗 메서드를 만드셨네요. 이 프라이빗 메서드는 없어도 괜찮을 것 같네요. 뭔가 다른 로직을 재사용한다던지, 아니면 추상화가 더 필요하다던지 다른 이유가 있으면 더 좋을 것 같아요

```suggestion
	public void send() throws IOException {
		httpExchange.sendResponseHeaders(statusCode, 0);
		httpExchange.close();
	}

	public void send(String responseBody) throws IOException {
		httpExchange.sendResponseHeaders(statusCode, responseBody.getBytes().length);
		httpExchange.getResponseBody().write(responseBody.getBytes());
		httpExchange.getResponseBody().close();
	}
```

EunjinJwa replied:
앗 불필요하게 작성되었네요^^; 제거했어요. 감사합니다

hannut91 reviewed (approved):
응답을 주는 일만 하는 객체를 만들어 주셨군요! 그래도 아직 static이 많긴한데, 이 부분은 스프링을 사용하면 줄어들 것도 많을 것 같아서 점차 적용하면 좋을 것 같습니다.

EunjinJwa committed:
CodeReview 반영

sendResponse 불필요한 메서드 제거
